convert2rhel/subscription.py

```diff
@@ -901,9 +916,8 @@ def download_rhsm_pkgs():
 
     if system_info.version.major == 7:
         pkgs_to_download += ["subscription-manager-rhsm", "python-syspurpose"]
-        _download_rhsm_pkgs(pkgs_to_download, _UBI_7_REPO_PATH, _UBI_7_REPO_CONTENT)
 
-    elif system_info.version.major == 8:
+    elif system_info.version.major >= 8:
         pkgs_to_download += [
             "python3-subscription-manager-rhsm",
             "dnf-plugin-subscription-manager",
```

Comment:
The package in el9 is named `libdnf-plugin-subscription-manager`.

```
$ sudo dnf --setopt=reposdir=/var/lib/convert2rhel/rhsm --releasever=9.2 --setopt=module_platform_id=platform:el9 repolist                              
repo id                                   repo name
ubi-9-baseos-convert2rhel                 Red Hat Universal Base Image 9 - BaseOS added by Convert2RHEL

$ sudo dnf --setopt=reposdir=/var/lib/convert2rhel/rhsm --releasever=9.2 --setopt=module_platform_id=platform:el9 search dnf-plugin-subscription-manager
Last metadata expiration check: 0:06:29 ago on Fri 11 Aug 2023 09:04:18 AM CDT.
============== Name Matched: dnf-plugin-subscription-manager ==============
libdnf-plugin-subscription-manager.x86_64 : Subscription Manager plugin for libdnf
``

Reply:
Also, python3-syspurpose doesn't exist in el9.

Reply:
Also, `subscription-manager-rhsm-certificates` matches two packages:

```
$ sudo rpm -qip *.rpm
Name        : subscription-manager-rhsm-certificates
Version     : 1.29.26.1
Release     : 1.el9_0
Architecture: x86_64
Install Date: (not installed)
Group       : Development/Libraries
Size        : 9716
License     : GPLv2
Signature   : RSA/SHA256, Tue 19 Jul 2022 08:59:04 AM CDT, Key ID 199e2f91fd431d51
Source RPM  : subscription-manager-1.29.26.1-1.el9_0.src.rpm
Build Date  : Mon 18 Jul 2022 02:21:29 PM CDT
Build Host  : x86-64-02.build.eng.rdu2.redhat.com
Packager    : Red Hat, Inc. <http://bugzilla.redhat.com/bugzilla>
Vendor      : Red Hat, Inc.
URL         : http://www.candlepinproject.org/
Summary     : Certificates required to communicate with a Red Hat Unified Entitlement Platform
Description :
This package contains certificates required for communicating with the REST interface
of a Red Hat Unified Entitlement Platform, used for the management of system entitlements
and to receive access to content.
Name        : subscription-manager-rhsm-certificates
Version     : 20220623
Release     : 1.el9
Architecture: noarch
Install Date: (not installed)
Group       : Unspecified
Size        : 27808
License     : GPLv2
Signature   : RSA/SHA256, Tue 28 Jun 2022 04:00:39 AM CDT, Key ID 199e2f91fd431d51
Source RPM  : subscription-manager-rhsm-certificates-20220623-1.el9.src.rpm
Build Date  : Thu 23 Jun 2022 08:07:45 AM CDT
Build Host  : ppc-002.build.eng.bos.redhat.com
Packager    : Red Hat, Inc. <http://bugzilla.redhat.com/bugzilla>
Vendor      : Red Hat, Inc.
URL         : https://www.candlepinproject.org/
Summary     : Certificates required to communicate with a Red Hat Unified Entitlement Platform
Description :
This package contains certificates required for communicating with the REST interface
of a Red Hat Unified Entitlement Platform, used for the management of system entitlements
and to receive access to content.
```

Comparing to a machine built originally with RHEL9 that I have, it looks like `subscription-manager-rhsm-certificates-20220623` might be the correct choice.  Forcing that worked.

Reply:
Thanks, I was looking for all RHEL 8-related things and changing them without looking into details. Let me change this one now.

Reply:
Once I hacked my way through these three things, I got a successful conversion from Alma 9 to RHEL 9.  So, it's close!

Reply:
The version of `subscription-manager-rhsm-certificates-20220623` is considered newer than the `subscription-manager-rhsm-certificates-1.29.26*` packages by dnf.

We're using `yumdownloader` which uses dnf under the hood, so it's going to pick up the right one when downloading it - the `20220623`.

Reply:
It does not pick up the right one, it gets both.

```
[2023-08-11T09:56:26-0500] TASK - [Convert: Subscription Manager - Download packages] ****************                                                                                                                                        
[2023-08-11T09:56:26-0500] DEBUG - Contents of /usr/share/convert2rhel/subscription-manager directory before RHEL rhsm packages download:                                                                                                     
                                                                                                                                                                                                                                              
[2023-08-11T09:56:26-0500] DEBUG - Downloading the subscription-manager package.                                                                                                                                                              
[2023-08-11T09:56:26-0500] DEBUG - Calling command 'yumdownloader -v --destdir=/usr/share/convert2rhel/subscription-manager --setopt=reposdir=/var/lib/convert2rhel/rhsm --releasever=9.2 --setopt=module_platform_id=platform:el9 subscriptio
n-manager'                                                                                                                                                                                                                                    
Successfully downloaded the subscription-manager package.                                                                                                                                                                                     
[2023-08-11T09:56:27-0500] DEBUG - Path of the downloaded package: /usr/share/convert2rhel/subscription-manager/subscription-manager-1.29.33.1-1.el9_2.x86_64.rpm                                                                             
[2023-08-11T09:56:27-0500] DEBUG - Downloading the subscription-manager-rhsm-certificates package.                                                                                                                                            
[2023-08-11T09:56:27-0500] DEBUG - Calling command 'yumdownloader -v --destdir=/usr/share/convert2rhel/subscription-manager --setopt=reposdir=/var/lib/convert2rhel/rhsm --releasever=9.2 --setopt=module_platform_id=platform:el9 subscriptio
n-manager-rhsm-certificates'                                                                                                                                                                                                                  
Successfully downloaded the subscription-manager-rhsm-certificates package.                                                                                                                                                                   
[2023-08-11T09:56:30-0500] DEBUG - Path of the downloaded package: /usr/share/convert2rhel/subscription-manager/subscription-manager-rhsm-certificates-1.29.26.1-1.el9_0.x86_64.rpm                                                           
[2023-08-11T09:56:30-0500] DEBUG - Downloading the python3-subscription-manager-rhsm package.                                                                                                                                                 
[2023-08-11T09:56:30-0500] DEBUG - Calling command 'yumdownloader -v --destdir=/usr/share/convert2rhel/subscription-manager --setopt=reposdir=/var/lib/convert2rhel/rhsm --releasever=9.2 --setopt=module_platform_id=platform:el9 python3-sub
scription-manager-rhsm'                                                                                                                                                                                                                       
Successfully downloaded the python3-subscription-manager-rhsm package.                                                                                                                                                                        
[2023-08-11T09:56:30-0500] DEBUG - Path of the downloaded package: /usr/share/convert2rhel/subscription-manager/python3-subscription-manager-rhsm-1.29.33.1-1.el9_2.x86_64.rpm                                                                
[2023-08-11T09:56:30-0500] DEBUG - Downloading the libdnf-plugin-subscription-manager package.                                                                                                                                                
[2023-08-11T09:56:30-0500] DEBUG - Calling command 'yumdownloader -v --destdir=/usr/share/convert2rhel/subscription-manager --setopt=reposdir=/var/lib/convert2rhel/rhsm --releasever=9.2 --setopt=module_platform_id=platform:el9 libdnf-plug
in-subscription-manager'                                                                                                                                                                                                                      
Successfully downloaded the libdnf-plugin-subscription-manager package.                                                                                                                                                                       
[2023-08-11T09:56:31-0500] DEBUG - Path of the downloaded package: /usr/share/convert2rhel/subscription-manager/libdnf-plugin-subscription-manager-1.29.33.1-1.el9_2.x86_64.rpm                                                               
[2023-08-11T09:56:31-0500] DEBUG - Downloading the python3-cloud-what package.                                                                                                                                                                
[2023-08-11T09:56:31-0500] DEBUG - Calling command 'yumdownloader -v --destdir=/usr/share/convert2rhel/subscription-manager --setopt=reposdir=/var/lib/convert2rhel/rhsm --releasever=9.2 --setopt=module_platform_id=platform:el9 python3-clo
ud-what'                                                                                                                                                                                                                                      
Successfully downloaded the python3-cloud-what package.                                                                                                                                                                                       
[2023-08-11T09:56:31-0500] DEBUG - Path of the downloaded package: /usr/share/convert2rhel/subscription-manager/python3-cloud-what-1.29.33.1-1.el9_2.x86_64.rpm                                                                               
[2023-08-11T09:56:31-0500] DEBUG - Downloading the json-c.x86_64 package.                                                                                                                                                                     
[2023-08-11T09:56:31-0500] DEBUG - Calling command 'yumdownloader -v --destdir=/usr/share/convert2rhel/subscription-manager --setopt=reposdir=/var/lib/convert2rhel/rhsm --releasever=9.2 --setopt=module_platform_id=platform:el9 json-c.x86_
64'                                                                                                                                                                                                                                           
Successfully downloaded the json-c.x86_64 package.                                                                                                                                                                                            
[2023-08-11T09:56:32-0500] DEBUG - Path of the downloaded package: /usr/share/convert2rhel/subscription-manager/json-c-0.14-11.el9.x86_64.rpm                                                                                                 
[2023-08-11T09:56:32-0500] DEBUG - Contents of /usr/share/convert2rhel/subscription-manager directory after RHEL rhsm packages download:                                                                                                      
subscription-manager-1.29.33.1-1.el9_2.x86_64.rpm                                                                                                                                                                                             
subscription-manager-rhsm-certificates-1.29.26.1-1.el9_0.x86_64.rpm                                                                                                                                                                           
subscription-manager-rhsm-certificates-20220623-1.el9.noarch.rpm                                                                                                                                                                              
python3-subscription-manager-rhsm-1.29.33.1-1.el9_2.x86_64.rpm                                                                                                                                                                                
libdnf-plugin-subscription-manager-1.29.33.1-1.el9_2.x86_64.rpm                                                                                                                                                                               
python3-cloud-what-1.29.33.1-1.el9_2.x86_64.rpm
json-c-0.14-11.el9.x86_64.rpm

[2023-08-11T09:56:32-0500] TASK - [Convert: Subscription Manager - Replace] **************************
Unregistering the system.
[2023-08-11T09:56:32-0500] DEBUG - Calling command 'rpm --quiet -q subscription-manager'
The subscription-manager package is not installed.
Removing installed subscription-manager/katello-ca-consumer packages.
No packages related to subscription-manager installed.
Checking for subscription-manager RPMs.
Installing subscription-manager RPMs.
[2023-08-11T09:56:32-0500] DEBUG - Calling command 'yum install -y --setopt=module_platform_id=platform:el9 /usr/share/convert2rhel/subscription-manager/subscription-manager-1.29.33.1-1.el9_2.x86_64.rpm /usr/share/convert2rhel/subscriptio
n-manager/subscription-manager-rhsm-certificates-1.29.26.1-1.el9_0.x86_64.rpm /usr/share/convert2rhel/subscription-manager/subscription-manager-rhsm-certificates-20220623-1.el9.noarch.rpm /usr/share/convert2rhel/subscription-manager/pytho
n3-subscription-manager-rhsm-1.29.33.1-1.el9_2.x86_64.rpm /usr/share/convert2rhel/subscription-manager/libdnf-plugin-subscription-manager-1.29.33.1-1.el9_2.x86_64.rpm /usr/share/convert2rhel/subscription-manager/python3-cloud-what-1.29.33
.1-1.el9_2.x86_64.rpm /usr/share/convert2rhel/subscription-manager/json-c-0.14-11.el9.x86_64.rpm'
Last metadata expiration check: 0:00:07 ago on Fri Aug 11 09:56:25 2023.
Package json-c-0.14-11.el9.x86_64 is already installed.
Error: 
 Problem: cannot install both subscription-manager-rhsm-certificates-20220623-1.el9.noarch and subscription-manager-rhsm-certificates-1.29.26.1-1.el9_0.x86_64
  - conflicting requests
(try to add '--allowerasing' to command line to replace conflicting packages or '--skip-broken' to skip uninstallable packages or '--nobest' to use not only best candidate packages)
CRITICAL - Failed to install subscription-manager packages. See the above yum output for details.
ERROR - (ERROR) PRE_SUBSCRIPTION.UNKNOWN_ERROR: Failed to install subscription-manager packages. See the above yum output for details.
```

```
$ sudo yumdownloader -v --destdir=/tmp/rpms --setopt=reposdir=/var/lib/convert2rhel/rhsm --releasever=9.2 --setopt=module_platform_id=platform:el9 subscription-manager-rhsm-certificates           
Loaded plugins: builddep, changelog, config-manager, copr, debug, debuginfo-install, download, generate_completion_cache, groups-manager, needs-restarting, playground, repoclosure, repodiff, repograph, repomanage, reposync, system-upgrade
DNF version: 4.14.0                                                                                                                                                                                                                           cachedir: /var/cache/dnf                                                                                               
User-Agent: constructed: 'libdnf (AlmaLinux 9.2; generic; Linux.x86_64)'
repo: using cache for: ubi-9-baseos-convert2rhel                                                                       
ubi-9-baseos-convert2rhel: using metadata from Tue 08 Aug 2023 10:35:13 AM CDT.
Last metadata expiration check: 0:05:33 ago on Fri 11 Aug 2023 09:56:26 AM CDT.                                                                                                                                                               
(1/2): subscription-manager-rhsm-certificates-20220623-1.el9.noarch.rpm                                 162 kB/s |  22 kB     00:00     
(2/2): subscription-manager-rhsm-certificates-1.29.26.1-1.el9_0.x86_64.rpm                               118 kB/s |  56 kB     00:00    
```

Reply:
Interesting, I'll ask around.

Reply:
Before I get the answer from the subscription-manager folks, I've specified that it should be the x86_64 one that's to be downloaded. That the arch that UBI 7 and 8 have and it was the easiest to do in the code for now. The only difference is that `20220623` one installs this one additional file: `/usr/share/licenses/subscription-manager-rhsm-certificates/COPYING`.